File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> prefix(nums.size(),1); 
        //1 2 3 4 
        //1 1 2 6
        for(int i = 1; i < prefix.size(); i++){
            prefix[i] = prefix[i-1] * nums[i-1]; 
        }
        int suffix = nums[nums.size()-1]; 
        for(int i = nums.size()-2; i >= 0; i--){
            prefix[i] *= suffix; 
            suffix *= nums[i]; 
        }
        return prefix; 
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length]; 
        answer[0] = 1; 
        for(int i = 1; i < nums.length; i++){
            answer[i] = nums[i-1] * answer[i-1]; 
        }

        int suffix = nums[nums.length-1]; 
        for(int i = nums.length-2; i >= 0; i--){
            answer[i] = suffix * answer[i]; 
            suffix *= nums[i]; 
        }

        return answer; 
    }
}